test/rulesets/Base/array.jl

```diff
@@ -171,7 +171,9 @@ end
 
         # Structured
         y, pb = rrule(reverse, Diagonal([1,2,3]))
-        @test unthunk(pb(rand(3,3))[2]) isa Diagonal
+        # We only preserve structure in this case if given structured tangent (no ProjectTo)
+        @test unthunk(pb(Diagonal([1.1, 2.1, 3.1]))[2]) isa Diagonal
+        @test unthunk(pb(rand(3, 3))[2]) isa AbstractArray
```

Comment:
Instead of just testing the absence of projection, how hard would it be to test something that doesn't want it?  Like `rrule(reverse, [(; x=[1,2]), (; x=[3,4])])`.

The structured arrays that do play well with reverse, and might reasonably show up in the wild, are row vectors:
```
julia> reverse([1,2,3]')
1×3 adjoint(::Vector{Int64}) with eltype Int64:
 3  2  1

julia> reverse([1,2,3]', dims=2)
1×3 adjoint(::Vector{Int64}) with eltype Int64:
 3  2  1
```
Not sure they need testing though, they will just work.

Reply:
Oh, too slow! Never mind.